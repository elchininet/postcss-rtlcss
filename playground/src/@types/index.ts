export type NumberOrString = string | number;
export type NumberOrStringArray = NumberOrString[];

export interface ObjectProps<T> {
    [key: string]: T;
}