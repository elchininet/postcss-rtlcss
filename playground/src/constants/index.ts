export enum COLORS {
    action = '#18AEC9',
    active = '#026C7F',
    gray = '#DDD',
    gray_dark = '#555',
    gray_darkest = '#1E1E1E',
    gray_light = '#EFEFEF',
    white = '#FFF'
}