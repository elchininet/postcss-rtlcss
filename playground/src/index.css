html,
body {
    height: 100%;
    margin: 0;
    padding: 0;
}

body {
    font-family: Arial, Helvetica, sans-serif;
    overscroll-behavior-y: contain;
}

#app {
    height: 100%;
}