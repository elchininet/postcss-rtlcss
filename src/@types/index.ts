import { Rule, AtRule } from 'postcss';

export interface ObjectWithProps<T> {
    [key: string]: T;
}

export enum Mode {
    combined = 'combined',
    override = 'override'
}

export enum Source {
    ltr = 'ltr',
    rtl = 'rtl'
}

export enum Autorename {
    disabled = 'disabled',
    flexible = 'flexible',
    strict = 'strict'
}

export type ModeValues = keyof typeof Mode;
export type SourceValues = keyof typeof Source;
export type AutorenameValues = keyof typeof Autorename;

export type strings = string | string[];

interface StringMapOptions {
    scope: '*' | 'selector' | 'url';
    ignoreCase: boolean;
}

export interface StringMap {
    name: string;
    priority: number;
    search: strings;
    replace: strings;
    options: StringMapOptions;
}

export interface PluginStringMap {
    name?: string;
    search: strings;
    replace: strings;
}

export interface PluginOptions {
    mode?: ModeValues;
    ltrPrefix?: strings;
    rtlPrefix?: strings;
    bothPrefix?: strings;
    source?: SourceValues;
    processUrls?: boolean;
    processKeyFrames?: boolean;
    useCalc?: boolean;
    stringMap?: PluginStringMap[];
    autoRename?: Autorename;
    greedy?: boolean;
}

export interface PluginOptionsNormalized extends Omit<Required<PluginOptions>, 'stringMap'> {
    stringMap: StringMap[];
}

export interface RulesObject {
    rule: Rule;
    ruleLTR: Rule | null;
    ruleRTL: Rule | null;
    ruleBoth: Rule | null;
}

export interface AtRulesObject {
    atRuleParams: string;
    atRule: AtRule;
    atRuleFlipped: AtRule;
}

export interface AtRulesStringMapObject {
    name: string;
    nameFlipped: string;
}

export interface AtRulesStringMap {
    [key: string]: AtRulesStringMapObject;
}

export interface ShortHandsData {
    [key: string]: {
        overridden: string | null;
        overrides: string[];
    };
}