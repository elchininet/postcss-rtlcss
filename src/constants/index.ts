export const COMMENT_TYPE = 'comment';
export const DECLARATION_TYPE = 'decl';
export const RULE_TYPE = 'rule';
export const AT_RULE_TYPE = 'atrule';
export const STRING_TYPE = 'string';
export const BOOLEAN_TYPE = 'boolean';
export const KEYFRAMES_NAME = 'keyframes';
export const ANIMATION_PROP = 'animation';
export const ANIMATION_NAME_PROP = 'animation-name';
export const RTL_COMMENT_REGEXP = /rtl:/;
export const RTL_CONTROL_DIRECTIVE_REG_EXP = /^\/\*!? *rtl:?(\w+)?:(\w+):?([^*]*?) *\*\/$/;
export const FLIP_PROPERTY_REGEXP = /(right|left)/i;

export enum CONTROL_DIRECTIVE {
    IGNORE = 'ignore'
}

export enum CONTROL_DIRECTIVE_BLOCK {
    BEGIN = 'begin',
    END = 'end'
}