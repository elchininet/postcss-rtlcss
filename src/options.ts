import { Mode, Source } from '@types';

export { Mode, Source };