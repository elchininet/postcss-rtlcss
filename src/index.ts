import postcss, { Root, Transformer } from 'postcss';
import { PluginOptions, RulesObject } from '@types';
import { parseOptions } from '@utilities/options';
import { parseRules } from '@rules/parser';
import { appendRules } from '@utilities/rules';
export { PluginOptions, Mode, Source, PluginStringMap,  } from '@types';

const transformer = (options: PluginOptions = {}): Transformer  => (
    (css: Root): void => {
        const optionsParsed = parseOptions(options);
        const result: RulesObject[] = parseRules(css, optionsParsed);
        appendRules(result);           
    }
);

export const postcssRTLCSS = postcss.plugin<PluginOptions>('postcss-rtlcss', transformer);