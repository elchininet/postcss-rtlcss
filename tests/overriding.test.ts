import postcss from 'postcss';
import postcssRTLCSS from '../src';
import { PluginOptions, Source } from '../src/@types';
import { readCSSFile, runTests } from './utils';

runTests({}, (pluginOptions: PluginOptions): void => {

  describe(`[[Mode: ${pluginOptions.mode}]] Overriding Tests: `, (): void => {

    let input = '';
  
    beforeEach(async (): Promise<void> => {
      input = input || await readCSSFile('overriding.css');
    });
  
    it('Basic', (): void => {
      const options: PluginOptions = { ...pluginOptions };
      const output = postcss([postcssRTLCSS(options)]).process(input);
      expect(output.css).toMatchSnapshot();
      expect(output.warnings()).toHaveLength(0);
    });
  
  });

});