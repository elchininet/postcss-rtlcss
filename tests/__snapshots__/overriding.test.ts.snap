// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[[Mode: combined]] Overriding Tests:  Basic 1`] = `
"/* Mirror properties, do nothing */
.test1 {
    padding-left: 10px;
    padding-right: 10px;
}

/* Mirror properties, do nothing */
.test2 {
    margin-left: 5px;
    margin-right: 5px;
    margin-left: 10px;
    margin-right: 10px;
    margin-left: 0;
    margin-right: 0;
}

/* Flipped-declarations overriden by non-flipped declarations, do nothing */
.test3 {
    text-align: left;
    text-align: right;
    text-align: center;
}

/* Overriden mirror properties */
.test4 {
    left: 0;
    right: 10px;
}

[dir=\\"ltr\\"] .test4 {
    left: 10px;
    right: 0;
}

[dir=\\"rtl\\"] .test4 {
    right: 10px;
    left: 0;
}"
`;

exports[`[[Mode: diff]] Overriding Tests:  Basic 1`] = `
".test4 {
    right: 10px;
    left: 0;
}"
`;

exports[`[[Mode: override]] Overriding Tests:  Basic 1`] = `
"/* Mirror properties, do nothing */
.test1 {
    padding-left: 10px;
    padding-right: 10px;
}

/* Mirror properties, do nothing */
.test2 {
    margin-left: 5px;
    margin-right: 5px;
    margin-left: 10px;
    margin-right: 10px;
    margin-left: 0;
    margin-right: 0;
}

/* Flipped-declarations overriden by non-flipped declarations, do nothing */
.test3 {
    text-align: left;
    text-align: right;
    text-align: center;
}

/* Overriden mirror properties */
.test4 {
    left: 0;
    left: 10px;
    right: 10px;
    right: 0;
}

[dir=\\"rtl\\"] .test4 {
    right: 10px;
    left: 0;
}"
`;
